codec-http2/src/test/java/io/netty/handler/codec/http2/DefaultHttp2ConnectionEncoderTest.java

```diff
@@ -346,19 +375,16 @@ public void headersWriteForUnknownStreamShouldCreateStream() throws Exception {
 
     @Test
     public void headersWriteShouldCreateHalfClosedStream() throws Exception {
-        int streamId = 5;
-        when(stream.id()).thenReturn(streamId);
-        when(stream.state()).thenReturn(IDLE);
-        mockFutureAddListener(true);
-        when(local.createStream(eq(streamId))).thenReturn(stream);
-        when(writer.writeHeaders(eq(ctx), eq(streamId), eq(EmptyHttp2Headers.INSTANCE), eq(0),
-                eq(DEFAULT_PRIORITY_WEIGHT), eq(false), eq(0), eq(true), eq(promise))).thenReturn(future);
-        encoder.writeHeaders(ctx, streamId, EmptyHttp2Headers.INSTANCE, 0, true, promise);
-        verify(local).createStream(eq(streamId));
-        verify(stream).open(eq(true));
-        // Trigger the write and mark the promise successful to trigger listeners
-        payloadCaptor.getValue().write(0);
-        promise.trySuccess();
+        Http2Connection connection = new DefaultHttp2Connection(true);
```

Comment:
I think it just has to do with granularity and isolation of tests.  The tests we are discussing have relatively "good" isolation and "low" granularity focused on their respective implementations they are designed to test.  The idea is that having this level of tests should result in a narrow set of locations where the problem may lie (hopefully not in the mock code :) ).  Not using mocks runs the risk of multiple pieces having bad behavior that may cancel each other out from the test's perspective.  We have tests at both levels (mocking everything around what we test, and just using real implementations) but generally the "real" implementation tests are relying on these "lower" level tests to catch "lower" level problems.

You bring up good points and these are often trade offs to consider while testing. The focus on the mock implementations should be the minimum as defined by the interface.  If the overhead and duplication is to burdensome or the interfaces are too flaky then we may want to re-asses.